src/connection.ts

```diff
@@ -3368,27 +3368,38 @@ Connection.prototype.STATE = {
               this.transitionTo(this.STATE.LOGGED_IN_SENDING_INITIAL_SQL);
             }
           } else if (this.ntlmpacket) {
-            const authentication = this.config.authentication as NtlmAuthentication;
-
-            const payload = new NTLMResponsePayload({
-              domain: authentication.options.domain,
-              userName: authentication.options.userName,
-              password: authentication.options.password,
-              ntlmpacket: this.ntlmpacket
-            });
-
-            this.messageIo.sendMessage(TYPE.NTLMAUTH_PKT, payload.data);
-            this.debug.payload(function() {
-              return payload.toString('  ');
-            });
-
-            this.ntlmpacket = undefined;
-
-            this.messageIo.readMessage().then((message) => {
-              this.dispatchEvent('message', message);
-            }, (err) => {
-              this.socketError(err);
-            });
+            try {
+              const authentication = this.config.authentication as NtlmAuthentication;
+
+              const payload = new NTLMResponsePayload({
+                domain: authentication.options.domain,
+                userName: authentication.options.userName,
+                password: authentication.options.password,
+                ntlmpacket: this.ntlmpacket
+              });
+
+              this.messageIo.sendMessage(TYPE.NTLMAUTH_PKT, payload.data);
+              this.debug.payload(function() {
+                return payload.toString('  ');
+              });
+
+              this.ntlmpacket = undefined;
+
+              this.messageIo.readMessage().then((message) => {
+                this.dispatchEvent('message', message);
+              }, (err) => {
+                this.socketError(err);
+              });
+            } catch (error: any) {
+              if (error.code === 'ERR_OSSL_EVP_UNSUPPORTED') {
+                this.emit('connect', new ConnectionError(`Node 17 now uses OpenSSL 3, which considers md4 encryption a legacy type.
+                In order to use NTLM with Node 17, enable the '--openssl-legacy-provider' command line flag.
+                Check the Tedious FAQ for more information.`, 'ELOGIN'));
+              } else {
+                this.emit('connect', new ConnectionError('Login failed.' + error.message, 'ELOGIN'));
```

Comment:
I think it would be best to just re-throw any "unknown" errors we encounter here. Otherwise other errors like syntax errors or trying to call methods on an `undefined` object and so on will get turned into `ConnectionErrors` here.

What do you think?